clang/labs/lab1/savotina_valeria/AddAttrAlwaysInline.cpp

```diff
@@ -0,0 +1,97 @@
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace {
+class AddAttrAlwaysInlineVisitor final
+    : public clang::RecursiveASTVisitor<AddAttrAlwaysInlineVisitor> {
+public:
+  explicit AddAttrAlwaysInlineVisitor(clang::ASTContext *context_)
+      : context(context_) {}
+  bool VisitFunctionDecl(clang::FunctionDecl *decl) {
+    if (decl->isFunctionOrFunctionTemplate()) {
+      if (!decl->hasAttr<clang::AlwaysInlineAttr>()) {
+        if (auto body = decl->getBody()) {
+          if (!findСonditionalStatement(body)) {
+            if (auto loc = decl->getSourceRange(); loc.isValid()) {
+              decl->addAttr(
+                  clang::AlwaysInlineAttr::Create(*context, loc.getBegin()));
+            }
+          }
+        }
+      }
+    }
```

Comment:
In LLVM it is customary to reduce nesting as much as possible:
```suggestion
    if (!decl->isFunctionOrFunctionTemplate())
      return true;
    if (decl->hasAttr<clang::AlwaysInlineAttr>())
      return true;
    if (auto body = decl->getBody()) {
      if (findСonditionalStatement(body))
        return true;

      if (auto loc = decl->getSourceRange(); loc.isValid())
        decl->addAttr(
            clang::AlwaysInlineAttr::Create(*context, loc.getBegin()));
    }
```

Reply:
@m-ly4 ok, I see, thanks!